pkg/sql/catalog/descs/collection.go

```diff
@@ -483,22 +483,49 @@ func (tc *Collection) getUserDefinedSchemaVersion(
 		return readFromStore()
 	}
 
-	// TODO (lucy): Replace this with a lookup to the parent database's Schema
-	// map to short-circuit returning negative results.
-	desc, shouldReadFromStore, err := tc.getLeasedDescriptorByName(
-		ctx, txn, dbID, keys.RootNamespaceID, schemaName)
+	// Look up whether the schema is on the database descriptor and return early
+	// if it's not.
+	// TODO (lucy): It's unfortunate that our current API (where we look up
+	// schemas by database ID and name) forces us to look up the leased database
+	// descriptor here, since we'll already have done this lookup when we're
+	// resolving the schema by name. Arguably we should be doing this at a higher
+	// level.
+	dbDesc, err := tc.GetDatabaseVersionByID(ctx, txn, dbID,
+		tree.DatabaseLookupFlags{CommonLookupFlags: tree.CommonLookupFlags{Required: true}})
 	if err != nil {
 		return nil, err
 	}
-	if shouldReadFromStore {
-		return readFromStore()
-	}
-	schema, ok := desc.(*schemadesc.Immutable)
-	if !ok {
+	schemaInfo, found := dbDesc.LookupSchema(schemaName)
+	if !found {
 		if flags.Required {
 			return nil, sqlerrors.NewUndefinedSchemaError(schemaName)
 		}
 		return nil, nil
+	} else if schemaInfo.Dropped {
+		if flags.Required {
+			return nil, pgerror.New(pgcode.InvalidSchemaName, "schema %s is being dropped")
+		}
+		return nil, nil
+	}
+
+	// If we have a schema ID from the database, get the schema descriptor. Since
```

Comment:
It doesn't seem like there is anything we can do about this without introducing that extra lookup right?

Reply:
Sorry, what do you mean by "this"?

Reply:
The false negatives

Reply:
Yeah, I don't think there's a way to completely eliminate this. We could maybe purge old leases for the database descriptor whenever we do it for the schema, but that still doesn't provide a guarantee.